LoRaEngine/LoraKeysManagerFacade/ConcentratorCredentialsFunction.cs

```diff
@@ -50,56 +53,62 @@ public class ConcentratorCredentialsFunction
             }
             catch (IncompatibleVersionException ex)
             {
-                log.LogError(ex, "Invalid version");
+                this.logger.LogError(ex, "Invalid version");
                 return new BadRequestObjectResult(ex.Message);
             }
 
-            return await RunFetchConcentratorCredentials(req, log, cancellationToken);
+            return await RunFetchConcentratorCredentials(req, cancellationToken);
         }
 
-        internal async Task<IActionResult> RunFetchConcentratorCredentials(HttpRequest req, ILogger log, CancellationToken cancellationToken)
+        internal async Task<IActionResult> RunFetchConcentratorCredentials(HttpRequest req, CancellationToken cancellationToken)
         {
             var stationEui = req.Query["StationEui"];
             if (StringValues.IsNullOrEmpty(stationEui))
             {
-                log.LogError("StationEui missing in request");
+                this.logger.LogError("StationEui missing in request");
                 return new BadRequestObjectResult("StationEui missing in request");
             }
 
             var credentialTypeQueryString = req.Query["CredentialType"];
             if (StringValues.IsNullOrEmpty(credentialTypeQueryString))
             {
-                log.LogError("CredentialType missing in request");
+                this.logger.LogError("CredentialType missing in request");
                 return new BadRequestObjectResult("CredentialType missing in request");
             }
             if (!Enum.TryParse<ConcentratorCredentialType>(credentialTypeQueryString.ToString(), out var credentialType))
             {
-                log.LogError("Could not parse '{QueryString}' to a ConcentratorCredentialType.", credentialTypeQueryString.ToString());
+                this.logger.LogError("Could not parse '{QueryString}' to a ConcentratorCredentialType.", credentialTypeQueryString.ToString());
                 return new BadRequestObjectResult($"Could not parse desired concentrator credential type '{credentialTypeQueryString}'.");
             }
 
             var twin = await this.registryManager.GetTwinAsync(stationEui, cancellationToken);
             if (twin != null)
             {
-                log.LogInformation("Retrieving '{CredentialType}' for '{StationEui}'.", credentialType.ToString(), stationEui.ToString());
+                this.logger.LogInformation("Retrieving '{CredentialType}' for '{StationEui}'.", credentialType.ToString(), stationEui.ToString());
                 try
                 {
-                    var cupsProperty = (string)twin.Properties.Desired[CupsPropertyName].ToString();
+                    const string cupsKey = "cups";
+                    if (!twin.Properties.Desired.TryReadJsonBlock(cupsKey, this.logger, out var cupsProperty))
+                        throw new ArgumentOutOfRangeException(cupsKey, "failed to parse cups config");
```

Comment:
[nit] why do we throw an `ArgumentOutOfRangeException` here? Maybe it would be a bit more accurate to mention that it's not the parsing that goes wrong, but more the retrieval of the JSON string.

Reply:
also this was changed.

Reply:
64d96951736fe0dc8930d143186c77566166178b